rfc/0009-policy-report.md

```diff
@@ -0,0 +1,323 @@
+```
+
+## Creating `ClusterPolicyReport` resource
+
+The `ClusterPolicyReport` object contains audit results for all the cluster wide
+Kubernetes resources.
+
+Since Kubewarden allows this kind of resources to be inspected only by
+`ClusterAdmissionPolicy` policies, the report will be built in the following way:
+
+* Find all the `ClusterAdmissionPolicy` policies that are interested about cluster wide
+  objects
+* Remove from this list all the `ClusterAdmissionPolicy` that are not interested
+  about `CREATE` events
+* Find all the cluster wide resources that are relevant to these policies
+* Evaluate the resources using the policies
+* Store the evaluation results inside of the `ClusterPolicyReport` object
+
+The CRD of `ClusterAdmissionPolicy` is basically equal to the one of `PolicyReport`.
+Because of that, we will use it in the same way as described in the previous
+section about `PolicyReport`.
+
+## Changes to `ClusterAdmissionPolicy` and `AdmissionPolicy`
+
+We will extend the `ClusterAdmissionPolicy` and `AdmissionPolicy` resources to have
+a new optional field called `background`.
+This field is going to hold a boolean value, which is going to be set to be `true` by default.
+
+Policies that have `background` set to false will be ignored during the background
+scans. Hence they will never show inside of the reports.
+
+# Drawbacks
+
+We will have just one `PolicyReport` per Namespace. Depending on the number of
+policies and Kubernetes resources defined inside of the Namespace, the size of
+the `policyReport.spec.results` could be significant.
+
+This could lead to `PolicyReport` objects too bit to be stored and/or degrade
+cluster performance because of the load put on ETCD.
+
+Quoting [this section](https://etcd.io/docs/v3.5/dev-guide/limit/#request-size-limit)
+of ETCD's documentation:
+
+>etcd is designed to handle small key value pairs typical for metadata. Larger requests will work, but may increase the latency of other requests. By default, the maximum size of any request is 1.5 MiB. This limit is configurable through --max-request-bytes flag for etcd server
+
+# Alternatives
+
+Instead of using the `PolicyReport` CRDs, we could adopt the same approach used
+by Gatekeeper [audit feature](https://open-policy-agent.github.io/gatekeeper/website/docs/audit/#constraint-status).
+
+Gatekeeper stores the report results inside of our equivalent of
+`AdmissionPolicy` and `ClusterAdmissionPolicy` objects.
+
+We could store the violation details inside of the `status` field of each
+`AdmissionPolicy` and `ClusterAdmissionPolicy`.
+
+Only the most recent audit runs are kept. There's also a limit on the number of
+detailed reports added to a policy. This is done to prevent the clogging of ETCD,
+as described in the "Drawbacks" section above.
+
+The main limitation about this approach is not going to be able to reuse the
+tool that is going to be built around the `PolicyReport` CRDs.
+
+# Unresolved questions
```

Comment:
Given that we have mutating policies, is there a possibility that several policies mutate a resource consecutively, getting trapped in an endless loop of mutations?

For example, these 2 policies may trigger each other in an endless loop.
- Policy A: mutates, adds `annot: prod`
- Policy B: mutates: if `annot: prod` present, change it to `annot: prod-touched`.

This may bring DOS, or crash of the cluster if the PolicyReports grow too large.